### ISSUE TITLE
Optimization is stuck when using new package versions

### ISSUE THREAD

**Issue by seadra** [SciML/DiffEqFlux.jl#875]
This code results works fine on a computer whose Julia packages are not up-to-date.
The optimizer finishes in a few seconds, with the final loss value in the order of 1e-7.

```julia
using DifferentialEquations, DiffEqFlux, Zygote, DiffEqSensitivity, Optimization, OptimizationFlux, OptimizationOptimJL, ComponentArrays, Lux, Random, LinearAlgebra

const T = 10.0;
const ω = π/T;

const id = Matrix{Complex{Float64}}(I,2, 2);
const u0 = id;


const utarget = Matrix{Complex{Float64}}([im 0; 0 -im]);

ann = Lux.Chain(Lux.Dense(1,32), Lux.Dense(32,32,tanh), Lux.Dense(32,1));
rng = Random.default_rng();
ip, st = Lux.setup(rng, ann);

function f_nn(u, p, t)
    local a, _ = ann([t/T],p,st);
    local A = [a[1] 0.0; 0.0 -a[1]];
    return -(im*A)*u;
end



tspan = (0.0, T)
prob_ode = ODEProblem(f_nn, u0, tspan, ComponentArray(ip));


function loss_adjoint(p)
    local prediction = solve(prob_ode, BS5(), p=p, abstol=1e-7, reltol=1e-7)
    local usol = last(prediction)
    local loss = abs(1.0 - abs(tr(usol*utarget')/2))
    return loss
end

opt_f = Optimization.OptimizationFunction((x, p) -> loss_adjoint(x), Optimization.AutoZygote());
opt_prob = Optimization.OptimizationProblem(opt_f, ComponentArray(ip));
optimized_sol_nn = Optimization.solve(opt_prob, AMSGrad(0.001), maxiters = 100, progress=true);
```

The list of packages on this computer is

```
(@v1.9) pkg> st
Status `~/.julia/environments/v1.9/Project.toml`
  [28f2ccd6] ApproxFun v0.13.24
  [6e4b80f9] BenchmarkTools v1.3.2
  [a134a8b2] BlackBoxOptim v0.6.2
  [d360d2e6] ChainRulesCore v1.18.0
⌅ [b0b7db55] ComponentArrays v0.13.7
  [8a292aeb] Cuba v2.3.0
  [acf6eb54] DFTK v0.6.11
  [717857b8] DSP v0.7.9
  [aae7a2af] DiffEqFlux v2.4.0
  [41bf760c] DiffEqSensitivity v6.79.0
⌃ [0c46a032] DifferentialEquations v7.6.0
⌃ [6a86dc24] FiniteDiff v2.17.0
⌅ [587475ba] Flux v0.13.13
  [f6369f11] ForwardDiff v0.10.36
  [a75be94c] GalacticOptim v3.4.0
  [14197337] GenericLinearAlgebra v0.3.11
  [f67ccb44] HDF5 v0.17.1
  [7073ff75] IJulia v1.24.2
  [de52edbc] Integrals v3.9.0
  [c31f79ba] IntegralsCubature v0.2.3
  [a98d9a8b] Interpolations v0.14.7
  [42fd0dbc] IterativeSolvers v0.9.3
  [0b1a1467] KrylovKit v0.6.0
  [b964fa9f] LaTeXStrings v1.3.0
⌅ [7ed4a6bd] LinearSolve v1.34.1
⌃ [bdcacae8] LoopVectorization v0.12.150
⌅ [b2108857] Lux v0.4.37
  [33e6dc65] MKL v0.6.1
  [eff96d63] Measurements v2.10.0
⌃ [961ee093] ModelingToolkit v8.36.0
  [bdf0d083] MultiFloats v1.0.4
  [76087f3c] NLopt v1.0.0
⌃ [315f7962] NeuralPDE v5.6.0
  [429524aa] Optim v1.7.8
⌃ [7f7a1694] Optimization v3.14.0
⌃ [3e6eede4] OptimizationBBO v0.1.4
⌃ [253f991c] OptimizationFlux v0.1.4
⌃ [36348300] OptimizationOptimJL v0.1.8
⌃ [1dea7af3] OrdinaryDiffEq v6.33.3
  [8314cec4] PGFPlotsX v1.6.0
  [5432bcbf] PaddedViews v0.5.12
  [ccf2f8ad] PlotThemes v3.1.0
  [91a5bcdd] Plots v1.39.0
  [8162dcfd] PrettyPrint v0.2.0
  [08abe8d2] PrettyTables v2.2.8
  [92933f4c] ProgressMeter v1.9.0
  [1fd47b50] QuadGK v2.9.1
  [67601950] Quadrature v2.1.0
  [88128e30] QuantumCircuitOpt v0.5.7
  [3c0b384b] QuantumInformation v0.4.10
⌃ [6e0679c1] QuantumOptics v1.0.8
  [f2b01f46] Roots v2.0.20
  [276daf66] SpecialFunctions v2.3.1
  [90137ffa] StaticArrays v1.6.5
⌃ [78aadeae] SymbolicNumericIntegration v1.0.0
⌅ [d1185830] SymbolicUtils v0.19.11
⌅ [0c5d862f] Symbolics v4.14.0
  [11d239b0] ThreadedIterables v0.2.1
⌃ [bc48ee85] Tullio v0.3.5
  [e88e6eb3] Zygote v0.6.67
  [9a3f8284] Random
```

The same code, when run on a computer with up-to-date packages gets stuck. The optimizer doesn't budge the loss value from its initial value, which is typically in the order of 0.1. The list of packages on that computer is

```
(@v1.9) pkg> st
Status `~/.julia/environments/v1.9/Project.toml`
  [28f2ccd6] ApproxFun v0.13.25
  [6e4b80f9] BenchmarkTools v1.3.2
  [b0b7db55] ComponentArrays v0.15.4
  [aae7a2af] DiffEqFlux v2.4.0
  [41bf760c] DiffEqSensitivity v6.79.0
⌃ [0c46a032] DifferentialEquations v7.7.0
⌅ [7da242da] Enzyme v0.10.18
⌅ [587475ba] Flux v0.13.17
  [7073ff75] IJulia v1.24.2
⌅ [de52edbc] Integrals v3.9.0
  [c31f79ba] IntegralsCubature v0.2.3
  [b964fa9f] LaTeXStrings v1.3.1
  [b2108857] Lux v0.5.10
  [33e6dc65] MKL v0.6.1
  [429524aa] Optim v1.7.8
⌃ [7f7a1694] Optimization v3.14.0
⌃ [3e6eede4] OptimizationBBO v0.1.4
⌃ [253f991c] OptimizationFlux v0.1.4
⌃ [36348300] OptimizationOptimJL v0.1.8
⌃ [42dfb2eb] OptimizationOptimisers v0.1.2
⌃ [1dea7af3] OrdinaryDiffEq v6.51.2
  [8314cec4] PGFPlotsX v1.6.0
  [5432bcbf] PaddedViews v0.5.12
  [ccf2f8ad] PlotThemes v3.1.0
  [91a5bcdd] Plots v1.39.0
  [92933f4c] ProgressMeter v1.9.0
⌃ [1ed8b502] SciMLSensitivity v7.29.0
  [de6bee2f] SimpleChains v0.4.6
  [276daf66] SpecialFunctions v2.3.1
  [90137ffa] StaticArrays v1.6.5
  [e88e6eb3] Zygote v0.6.67
  [37e2e46d] LinearAlgebra
  [9a3f8284] Random
```

Both computers have the same OS environment otherwise, both Intel, with Julia 1.9.3.

**Comment by ChrisRackauckas** [comment]
Is there a reason you're still using the deprecated DiffEqSensitivity instead of SciMLSensitivity?

**Comment by seadra** [comment]
Ah, I just haven't kept up with the changes in the SciML it seems, thanks for letting me know!

I removed DiffEqSensitivity, installed SciMLSensitivity, and replaced DiffEqSensitivity with SciMLSensitivity in the `using ...` line. Running the code, now I get the following error:

```
┌ Warning: Reverse-Mode AD VJP choices all failed. Falling back to numerical VJPs
└ @ SciMLSensitivity ~/.julia/packages/SciMLSensitivity/rYAAz/src/concrete_solve.jl:92

MethodError: no method matching default_relstep(::Nothing, ::Type{ComplexF64})

Closest candidates are:
  default_relstep(::Type, ::Any)
   @ FiniteDiff ~/.julia/packages/FiniteDiff/40JnL/src/epsilons.jl:25
  default_relstep(::Val{fdtype}, ::Type{T}) where {fdtype, T<:Number}
   @ FiniteDiff ~/.julia/packages/FiniteDiff/40JnL/src/epsilons.jl:26
```

**Comment by ChrisRackauckas** [comment]
It looks like this is covered by the other issue then as one just about complex numbers?

**Comment by seadra** [comment]
Unfortunately, no. The other issue is about the compilation failing with SciMLSensitivity, whereas in this issue, the code compiles and run (with DiffEqSensitivity) but the optimiziation results are bad.

So can we re-open this?

**Comment by ChrisRackauckas** [comment]
What's the issue with SciMLSensitivity? There is no DiffEqSensitivity to be updated. Can you share a SciMLSensitivity version of this?

**Comment by seadra** [comment]
The SciMLSensitivity version gives the error [above](https://github.com/SciML/DiffEqFlux.jl/issues/875#issuecomment-1812673018), and the full code is in the first post of https://github.com/SciML/SciMLSensitivity.jl/issues/938

There is no substantial change between the two actually. I just replaced DiffEqSensitivity with SciMLSensitivity in the using statement.

**Comment by seadra** [comment]
I didn't claim this current issue is caused by DiffEqSensitivity, by the way. It is possibly caused by something else.

DiffEqSensitivity only came up when you asked why I import a deprecated package, but I think the fact that the identical code optimizes fine on one machine and not on the other machine, both importing DiffEqSensitivity points to something else.

**Comment by seadra** [comment]
Since it's not obvious this is a DiffEqSensitivity issue (I'd love to explicitly rule it out once SciMLSensitivity issue with complex numbers is fixed), can we keep this issue open?

**Comment by ChrisRackauckas** [comment]
DiffEqSensitivity was deprecated into SciMLSensitivity over a year and a half ago and the only thing that exists is old versions of it in the SciMLSensitivity repository. There is no DiffEqSensitivity.jl to be updated or patched. 

> I didn't claim this current issue is caused by DiffEqSensitivity, by the way. It is possibly caused by something else.

> Since it's not obvious this is a DiffEqSensitivity issue (I'd love to explicitly rule it out once SciMLSensitivity issue with complex numbers is fixed), can we keep this issue open?

Why would this not be a DiffEqSensitivity.jl issue? There is nothing in your code that even uses DiffEqFlux. You can remove it from your using statements and it wouldn't matter. The only thing it could be is DiffEqSensitivity. I don't understand what the connection to DiffEqFlux is given that none of its layers are used.

**Comment by seadra** [comment]
> Why would this not be a DiffEqSensitivity.jl issue?

Because, as I mentioned, everything runs fine using the identical version of DiffEqSensitivity on this current computer. 

If it is a DiffEqSensitivity issue, then I don't understand why I can optimize things perfectly with the code at the top on this computer.

**Comment by ChrisRackauckas** [comment]
Going through these old versions won't be productive. If I had to guess what it was, there was something that happened when we introduced Enzyme defaults about 1 1/4 years ago which was fixed by the v0.11 update, which is why we had a short period of defaulting to Enzyme before pulling that back, and then fixed it up, relating to the choice of DuplicatedNoNeed. Given that I see "[7da242da] Enzyme v0.10.18", I'd guess it's that behavior around Enzyme defaults which could have caused it, or some rule that moved around in Zygote. If you're willing to do the archeology that's fine, but it's almost certainly something related to that derivative overload handling that got fixed, or its something that we turned into an error to ensure we wouldn't get false zeros from Zygote, in which case "Warning: Reverse-Mode AD VJP choices all failed. Falling back to numerical VJPs" is from the new Zygote errors we added to improve safety telling the default algorithm that it shouldn't use Zygote here. 

Either way, the fix is to investigate the SciMLSensitivity issue and get the real case fixed up here.

**Comment by seadra** [comment]
I did some monkey patching, and one of the things I already tried was to remove Enzyme (among other packages), but it unfortunately didn't fix the problem on the computer.

Indeed, since there are many possibilities, and your best guess seems to be DiffEqSensitivity, I'll come back to this issue once SciMLSensitivity works with loss functions that use complex matrices during intermediate calculations.
